feat(image_optimizer_default_settings): add Image Optimizer default settings API

daboross opened the pull request:
Includes an update to go-fastly 9.4.0.

daboross committed:
Add Image Optimizer default settings API

daboross committed:
Temporarily vendor go-fastly from Image Optimizer default settings API PR

daboross committed:
Generate docs.

Integralist commented on fastly/block_fastly_service_image_optimizer_default_settings.go:
```diff
@@ -0,0 +1,208 @@
+			apiInput.Upscale = &upscale
+		case "allow_video":
+			allowVideo := value.(bool)
+			apiInput.AllowVideo = &allowVideo
```
Everywhere you've created a variable for the sake of taking its address you can instead use `gofastly.ToPointer(value.(whatever))`.

daboross replied:
Oh awesome! That's much better haha.

Integralist reviewed (commented):
Just one small comment for the moment, but it's looking good. Let me know when you're ready for a full review.

daboross committed:
Handle case where reading image optimizer settings on a service without IO enabled.

daboross committed:
Fix "invalid key for element" error caused by missing "name" key

daboross committed:
PR cleanup suggestion.

daboross committed:
Include default values.

I was initially trying to get Terraform to ignore settings that wern't
set in the configuration, leaving them at their previously configured
value. This seems to be counter to Terraform's design, though, so I've
relented & included the current default values for each setting here.

It seems like maybe we could have done that for the string values, and
non-0 integers, but it seems like a nonstarter for booleans
(https://github.com/hashicorp/terraform-plugin-sdk/issues/817), and
I didn't want to go against Terraform's design too much.

daboross committed:
Update go-fastly vendored from PR

daboross committed:
Fix lints + add debug message.

daboross committed:
One more vendor update.

daboross committed:
Update jpeg_type, and with that, finally have a passing test!

daboross commented:
@Integralist I'd request a full review whenever you've got the time!

I'll leave this as a draft until the api-documentation & go-fastly PRs are merged, but I've finally gotten this working.

Also, my apologies, I made some modifications to the go-fastly PR to better test it & better support the terraform use case. Probably should have waited until I had everything working from one end to the other to submit PRs, but well you live & learn.

I'll clean up / squash the commit messages - they're a bit of a mess - but the code itself should be good.

daboross committed:
Update vendored code again.

daboross committed:
Update field documentation.

Part of this is updating field with descriptions from the final
api-documentation PR. The other part is just my messing with stuff to
try to make it look better.

daboross committed:
Update generated docs

(also, move image_optimizer_default_settings guide from docs to templates)

daboross committed:
Change Delete so it resets all settings to default.

Integralist commented:
👋🏻 @daboross apologies I've been stacked up with other work and I didn't realise it's been 2 weeks since you last commented. So I'll try to take a look at this PR tomorrow morning for you.

daboross commented:
:wave:

It's no problem! I also could have commented something: after finalizing the api-documentation PR, I went and cleaned up the go-fastly PR, and was starting to clean up this one up when I got sidetracked by a missing consideration in the original API.

I'll push those commits so tomorrow morning you can check it out, and leave a comment where I have one part untested because of a change needed to the API itself.

daboross commented on fastly/block_fastly_service_image_optimizer_default_settings.go:
```diff
@@ -0,0 +1,336 @@
+		if he, ok := err.(*gofastly.HTTPError); ok {
+			if he.StatusCode == http.StatusBadRequest {
+				for _, e := range he.Errors {
+					if strings.Contains(e.Detail, "Image Optimizer is not enabled on this service") {
```
This is the untested bit! When editing the docs for this PR, I was thinking about whether we needed a warning for this API requiring customers to be entitled, and realized we never thought about nor implemented any checks for that.

So I'm working on a PR to add a requirement that Image Optimizer is enabled on the service, and I'm waiting to merge the api-documentation PR (and to un-draft these) until that's ready.

daboross replied:
I've now tested that this works.

Integralist commented on fastly/block_fastly_service_image_optimizer_default_settings_test.go:
```diff
@@ -0,0 +1,97 @@
+					testAccCheckServiceExists("fastly_service_vcl.foo", &service),
+					resource.TestCheckResourceAttr("fastly_service_vcl.foo", "name", serviceName),
+					resource.TestCheckResourceAttr("fastly_service_vcl.foo", "backend.#", "1"),
+					testAccCheckFastlyServiceVCLBackendAttributes(&service, []*gofastly.Backend{&b1}),
```
We should have some test assertions for validating the `image_optimizer_default_settings` block was created correctly in the TF state, and also add an 'update' flow so we can be sure it updates correctly too.

daboross replied:
Thanks for commenting on this! I've added a validation check, and a second step for checking updating.

Integralist reviewed (commented):
LGTM, just one comment re: tests...

daboross committed:
Overhaul Image Optimizer default settings test.

daboross commented:
Going to rebase to fix merge conflicts, keeping all commits as is otherwise.

Integralist reviewed (approved):
LGTM. Mark the PR 'ready for review' and I'll give it a final going over.

daboross committed:
Update vendored go-fastly to 9.4.0.

daboross committed:
Add debug message for ignoring image-optimizer-disabled error on deletion.

This is working well! Just wanted to add this to double-confirm it's
doing what I expect.

daboross committed:
Remove toolchain directive.

daboross commented:
This should be ready for a full review! I've included an update to the released go-fastly 9.4.0.

daboross committed:
Undo go version bump.

Not sure why updating had caused this, but this should still work.
